samza-core/src/main/java/org/apache/samza/task/SystemTimerSchedulerFactory.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.task;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Per-task factory for keyed timer schedulers.
+ */
+public class SystemTimerSchedulerFactory {
+
+  /**
+   * For run loop to listen to timer firing so it can schedule the callbacks.
+   */
+  public interface TimerListener {
+    void onTimer();
+  }
+
+  private final ScheduledExecutorService executor;
+  private final Map<Object, SystemTimerScheduler> schedulers = new HashMap<>();
+  private final Map<TimerKey<?>, TimerCallback> readyTimers = new ConcurrentHashMap<>();
+  private TimerListener timerListener;
+
+  public static SystemTimerSchedulerFactory create(ScheduledExecutorService executor) {
+    return new SystemTimerSchedulerFactory(executor);
+  }
+
+  private SystemTimerSchedulerFactory(ScheduledExecutorService executor) {
+    this.executor = executor;
+  }
+
+  public <K> SystemTimerScheduler<K> getScheduler(K key) {
+    SystemTimerScheduler<K> scheduler = schedulers.get(key);
+    if (scheduler == null) {
+      scheduler = new SystemTimerScheduler<>(key);
+      schedulers.put(key, scheduler);
+    }
+    return scheduler;
+  }
+
+  public <K> void removeScheduler(K key) {
+    final SystemTimerScheduler<K> scheduler = schedulers.get(key);
+    if (scheduler != null) {
+      scheduler.cancel();
+      schedulers.remove(key);
+    }
+  }
+
+  void registerListener(TimerListener listener) {
+    timerListener = listener;
+  }
+
+  public Map<TimerKey<?>, TimerCallback> removeReadyTimers() {
+    final Map<TimerKey<?>, TimerCallback> timers = new TreeMap<>(readyTimers);
+    readyTimers.keySet().removeAll(timers.keySet());
+    return timers;
+  }
+
+  public class SystemTimerScheduler<K> {
```

Comment:
This is not really a scheduler, right? It's just holding the key and the future for the scheduled task. Also, since this is not a Scheduler, we don't need a SchedulerFactory, or a per key "scheduler". Can we clean up these interfaces/naming?

Reply:
Agree that the API should be typed, but the container doesn't currently care about the type of the key. Even for High Level API/BEAM, I'd think that the type would only matter downstream of StreamOperatorTask (e.g. in WindowImpl for HLA). 

Unless there's some type-checking reasons it's required, maybe we can go with the simpler implementation first, and make it typed when the framework actually needs to know about the type of the keys? If we do that, a simple per task Object -> Future map would probably suffice.

Reply:
Sure, I can rename this to be a timer holder, and move the schedule function outside.

Reply:
Fixed by changing the map to be just Map<Object, SchedulerFuture>.